AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy

```diff
@@ -266,10 +267,65 @@ module AwsCryptographyMaterialProvidersOperations refines AbstractAwsCryptograph
     //= type=implication
     //# If no max cache size is provided, the crypotgraphic materials cache MUST be configured to a
     //# max cache size of 1000.
-    var cache := if input.cache.Some? then
-      input.cache.value
-    else
-      Types.Default(Types.DefaultCache(entryCapacity := 1000));
+    var cmc;
+
+    if input.cache.Some? {
+      match input.cache.value {
+        case Shared(c) =>
+          cmc := c;
+        case _ =>
+          cmc :- CreateCryptographicMaterialsCache(
+            config,
+            CreateCryptographicMaterialsCacheInput(cache := input.cache.value)
+          );
+      }
+    }
+    else {
+      cmc :- CreateCryptographicMaterialsCache(
+        config,
+        CreateCryptographicMaterialsCacheInput(
+          cache := Types.Default(
+            Types.DefaultCache(entryCapacity := 1000)
+          )
+        )
+      );
+    }
+
+    // Set the correct partitionId
+    // If the partitionID is provided, convert it into UTF8 bytes
+    // If partitionID is not provided, generate a random UUID
+    var partitionIdBytes : seq<uint8>;
+
+    if input.partitionId.Some? {
+      partitionIdBytes :- UTF8.Encode(input.partitionId.value)
+      .MapFailure(
+        e => Types.AwsCryptographicMaterialProvidersException(
+            message := "Could not UTF-8 Encode Partition ID: " + e
+          )
+      );
+    } else {
+      var maybeUuid := UUID.GenerateUUID();
+
+      var uuid :- maybeUuid
+      .MapFailure(e => Types.AwsCryptographicMaterialProvidersException(message := e));
```

Comment:
@seebees has turned me to the darkside and I now prefer these types of statements like, 
```
var uuid? := ...;
var uuid :- uuid?.MapFailure...
```

Reply:
Updated. Thanks for the suggestion.